text/000-rustdoc-jump-to-definition.md

```diff
@@ -73,20 +71,117 @@ Another important note is that if documentation is split in multiple places or r
 
+Here is a video showing the feature in action:
+
+https://user-images.githubusercontent.com/3050060/114622354-21307b00-9cae-11eb-834d-f6d8178a37bd.mp4
+
+
 # Reference-level explanation
 [reference-level-explanation]: #reference-level-explanation
 
-It doesn't require javascript since it's only generating links, making it work with or without javascript enabled.
+This feature **only** targets the source code pages.
+
+It can be implemented without JavaScript additions to Rustdoc as pure link generation; so it will work without JS enabled. Rustdoc already has an internal link generation system that can be used here.
+
+## Link generation
+
+There are two kinds of links:
+
+ * Jump to documentation.
+ * Jump to definition.
+
+### Jump to documentation
+
+All items with a documentation page should have a link to it (like methods, trait/impl associated items, etc). However, fields, variants and impl blocks should not (explications below).
+
+The link is generated on their ident/name where they are declared, not when they are in use. For example:
+
+```rust
+pub struct Foo;
+
+let x = Foo;
+```
+
+On `Foo` definition (`pub struct Foo`), we generate a link to its documentation page whereas on its usage (`x = Foo`), we generate a link to jump to its definition.
+
+This is why generating a link for the impl blocks would be complicated: on `impl Foo`, we generate a "jump to definition" link on `Foo`, leaving no space to generate a link to the impl documentation.
+
+For fields and variants, it's simply to limit the quantity of links generated: you can simply click on their parent item to get to their documentation page.
+
+### Jump to definition
 
-For the link generation, we use the already existing system in rustdoc. Meaning their generation will follow the same rules as the link in the documentation pages.
+A small preamble before listing the items we should generate links for: if an item which should get a link is preceded by a path, the whole path is used as a link. For example:
 
-All items with documentation should have a link to their documentation page (like impl blocks, methods, trait/impl associated items, etc).
+```rust
+let x: b::c::D; // `b::c::D` is the link to `D`.
+foo::some_fn(); // `foo::some_fn` is the link to `some_fn`.
+```
+
+Anything not specifically mentioned should not get a link generated for itself (such as variable name):
+
+```rust
+fn f(mut a: String) {
+    a.push_str("something"); // `a` should not get a link but `push_str` should!
```

Comment:
We should probably have an explicit section about "local variables" saying that they do _not_ JTD

Reply:
That would make it more logical considering there are two sections for primitive types and fields.

Reply:
Under the current implementation, local variables _do_ get a JTD link. For instance in https://rustdoc.crud.net/imperio/jump-to-def-regex/src/regex/pikevm.rs.html#205, at_next links to a `let` statement. Either the implementation is incorrect or the description here in the RFC that "`a` should not get a link" is incorrect.

Reply:
The implementation is incorrect. A few things are still "rough" and need some polishing. But we also all need to agree on what should or should not get links. Personally I think it's too much for variables but I'm curious to see if others share the opposite point of view.

Reply:
I am in favor of supporting local variables. Here's why:

- We should definitely linkify global variables, because they may be far away or in another file.
- If you're looking at a variable reference in a function, you don't know whether it's a local variable or a global one (or one from an enclosing scope). By linkifying variables we make it easy for people to find out the answer. If we linkify only some variables (the global ones), we will leave people scratching their head about why some things are linkified and some are not.

Reply:
This is a fair argument. Should I put this argument directly in the section for local variables too? Having arguments alongside such cases could be useful but not sure if it's supposed to be in the RFC directly.

Reply:
Yes, I think it makes sense to put the argument in the local variable section for now. It may make sense to reorganize it elsewhere based on subsequent edits.